**Hacker News data set analysis**  

We analyze the data set of Hacker News technology site to see which topics get the most comments and at which creation date and other aspects.

In [1]:
import csv as c

In [4]:
opened_file = open('hacker_news.csv')
read_file = c.reader(opened_file)
hn = list(read_file)
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [5]:
headers = hn[0]
hn = hn[1:]
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [7]:
ask_posts = []
show_posts = []
other_posts = [] 

In [8]:
for row in hn:
    title = row[1].lower()
    if title.startswith("ask hn"):
        ask_posts.append(row)
    elif title.startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))
        

1744
1162
17194


So as we want to analyze ask Hacker News and show Hacker News posts only, we separated them into two list of lists,the first category contains 1744 posts while the latter contains 1162 post.

In [9]:
total_ask_comments = 0
for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments
avg_ask_comments = total_ask_comments/len(ask_posts)
print(avg_ask_comments)

14.038417431192661


In [10]:
total_show_comments = 0
for row in show_posts:
    num_comments = int(row[4])
    total_show_comments += num_comments
avg_show_comments = total_show_comments/len(show_posts)
print(avg_show_comments)

10.31669535283993


We find out that ask posts receive on average 14 comments for a post which is more than the show posts where a post receives only 10 comments on average.

Since the ask posts receive more comments we will analyze these posts to know which hour during the day the posts created at would receive more comments.

In [11]:
import datetime as dt

In [16]:
result_list = []
for row in ask_posts:
    date_comments=[row[6],int(row[4])]
    result_list.append(date_comments)
print(result_list[:5])    

[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 3], ['10/15/2015 16:38', 17]]


In [20]:
counts_by_hour = {}
comments_by_hour = {}
for row in result_list:
    created_at = dt.datetime.strptime(row[0],"%m/%d/%Y\n%H:%M")
    hour = created_at.strftime("%H")
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = row[1]
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += row[1]

In [22]:
print(counts_by_hour)
print(comments_by_hour)

{'14': 107, '07': 34, '08': 48, '04': 47, '13': 85, '05': 46, '18': 109, '19': 110, '03': 54, '23': 68, '16': 108, '22': 71, '10': 59, '00': 55, '02': 58, '15': 116, '20': 80, '01': 60, '09': 45, '17': 100, '11': 58, '12': 73, '21': 109, '06': 44}
{'14': 1416, '07': 267, '08': 492, '04': 337, '13': 1253, '05': 464, '18': 1439, '19': 1188, '03': 421, '23': 543, '16': 1814, '22': 479, '10': 793, '00': 447, '02': 1381, '15': 4477, '20': 1722, '01': 683, '09': 251, '17': 1146, '11': 641, '12': 687, '21': 1745, '06': 397}


In [33]:
avg_by_hour = []
for i in comments_by_hour:
    hour_average = [i,(comments_by_hour[i]/counts_by_hour[i])]
    avg_by_hour.append(hour_average)
print(avg_by_hour)    
    

[['14', 13.233644859813085], ['07', 7.852941176470588], ['08', 10.25], ['04', 7.170212765957447], ['13', 14.741176470588234], ['05', 10.08695652173913], ['18', 13.20183486238532], ['19', 10.8], ['03', 7.796296296296297], ['23', 7.985294117647059], ['16', 16.796296296296298], ['22', 6.746478873239437], ['10', 13.440677966101696], ['00', 8.127272727272727], ['02', 23.810344827586206], ['15', 38.5948275862069], ['20', 21.525], ['01', 11.383333333333333], ['09', 5.5777777777777775], ['17', 11.46], ['11', 11.051724137931034], ['12', 9.41095890410959], ['21', 16.009174311926607], ['06', 9.022727272727273]]


In [34]:
swap_avg_by_hour = []
for i in avg_by_hour:
    swap = [i[1],i[0]]
    swap_avg_by_hour.append(swap)
print(swap_avg_by_hour)    

[[13.233644859813085, '14'], [7.852941176470588, '07'], [10.25, '08'], [7.170212765957447, '04'], [14.741176470588234, '13'], [10.08695652173913, '05'], [13.20183486238532, '18'], [10.8, '19'], [7.796296296296297, '03'], [7.985294117647059, '23'], [16.796296296296298, '16'], [6.746478873239437, '22'], [13.440677966101696, '10'], [8.127272727272727, '00'], [23.810344827586206, '02'], [38.5948275862069, '15'], [21.525, '20'], [11.383333333333333, '01'], [5.5777777777777775, '09'], [11.46, '17'], [11.051724137931034, '11'], [9.41095890410959, '12'], [16.009174311926607, '21'], [9.022727272727273, '06']]


In [35]:
sorted_swap = sorted(swap_avg_by_hour,reverse=True)

In [37]:
Top5 = "Top 5 Hours for Ask Posts Comments:"
print(Top5)
for i,j in sorted_swap[:5]:
    hour = dt.datetime.strptime(j,"%H")
    hour = hour.strftime("%H:%M")
    string = "{}:{:.2f} average comments per post.".format(hour,i)
    print(string)
    

Top 5 Hours for Ask Posts Comments:
15:00:38.59 average comments per post.
02:00:23.81 average comments per post.
20:00:21.52 average comments per post.
16:00:16.80 average comments per post.
21:00:16.01 average comments per post.


We found the count of ask posts and the number of comments for those posts that are created at each hour during the day. Then we could find the average comments for posts created at each hour, and after sorting the results we were able to find the top five hours during the day that receive most comments based on Hacker News site ask posts data.